In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [2]:
dataset = pd.read_csv("Employee_Attrition_Dataset.csv")
dataset

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [3]:
dataset.drop(['EmployeeCount', 'EmployeeNumber','StandardHours'], axis = 1, inplace = True)

In [4]:
df1 = pd.get_dummies(dataset, dtype = int, drop_first = True)

#### Checking for Null Values

In [5]:
dataset.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [6]:
df1.columns

Index(['Age', 'DailyRate', 'DistanceFromHome', 'Education',
       'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel',
       'JobSatisfaction', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
       'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole',
       'YearsSinceLastPromotion', 'YearsWithCurrManager', 'Attrition_Yes',
       'BusinessTravel_Travel_Frequently', 'BusinessTravel_Travel_Rarely',
       'Department_Research & Development', 'Department_Sales',
       'EducationField_Life Sciences', 'EducationField_Marketing',
       'EducationField_Medical', 'EducationField_Other',
       'EducationField_Technical Degree', 'Gender_Male',
       'JobRole_Human Resources', 'JobRole_Laboratory Technician',
       'JobRole_Manager', 'JobRole_Manufacturing Director',
       'JobRole_Research Director', 'JobRole

# Univariate Analysis

In [7]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EnvironmentSatisfaction   1470 non-null   int64 
 9   Gender                    1470 non-null   object
 10  HourlyRate                1470 non-null   int64 
 11  JobInvolvement            1470 non-null   int64 
 12  JobLevel                  1470 non-null   int64 
 13  JobRole                   1470 non-null   object
 14  JobSatisfaction         

In [8]:
def qualQuan(dataset):
    quan = []
    qual = []
    for columnname in dataset.columns:
        if dataset[columnname].dtype == 'O':
            qual.append(columnname)
        else:
            quan.append(columnname)
    return quan, qual

In [9]:
quan, qual = qualQuan(dataset)

In [10]:
quan

['Age',
 'DailyRate',
 'DistanceFromHome',
 'Education',
 'EnvironmentSatisfaction',
 'HourlyRate',
 'JobInvolvement',
 'JobLevel',
 'JobSatisfaction',
 'MonthlyIncome',
 'MonthlyRate',
 'NumCompaniesWorked',
 'PercentSalaryHike',
 'PerformanceRating',
 'RelationshipSatisfaction',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'WorkLifeBalance',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [11]:
qual

['Attrition',
 'BusinessTravel',
 'Department',
 'EducationField',
 'Gender',
 'JobRole',
 'MaritalStatus',
 'Over18',
 'OverTime']

In [12]:
df1['Attrition_Yes'].value_counts()

Attrition_Yes
0    1233
1     237
Name: count, dtype: int64

In [13]:
dataset.describe()

Age    DailyRate  DistanceFromHome    Education  \
count  1470.000000  1470.000000       1470.000000  1470.000000   
mean     36.923810   802.485714          9.192517     2.912925   
std       9.135373   403.509100          8.106864     1.024165   
min      18.000000   102.000000          1.000000     1.000000   
25%      30.000000   465.000000          2.000000     2.000000   
50%      36.000000   802.000000          7.000000     3.000000   
75%      43.000000  1157.000000         14.000000     4.000000   
max      60.000000  1499.000000         29.000000     5.000000   

       EnvironmentSatisfaction   HourlyRate  JobInvolvement     JobLevel  \
count              1470.000000  1470.000000     1470.000000  1470.000000   
mean                  2.721769    65.891156        2.729932     2.063946   
std                   1.093082    20.329428        0.711561     1.106940   
min                   1.000000    30.000000        1.000000     1.000000   
25%                   2.000000    48.000000        2.000000     1.000000   
50%                   3.000000    66.000000        3.000000     2.000000   
75%                   4.000000    83.750000        3.000000     3.000000   
max                   4.000000   100.000000        4.000000     5.000000   

       JobSatisfaction  MonthlyIncome  ...  PerformanceRating  \
count      1470.000000    1470.000000  ...        1470.000000   
mean          2.728571    6502.931293  ...           3.153741   
std           1.102846    4707.956783  ...           0.360824   
min           1.000000    1009.000000  ...           3.000000   
25%           2.000000    2911.000000  ...           3.000000   
50%           3.000000    4919.000000  ...           3.000000   
75%           4.000000    8379.000000  ...           3.000000   
max           4.000000   19999.000000  ...           4.000000   

       RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
count               1470.000000       1470.000000        1470.000000   
mean                   2.712245          0.793878          11.279592   
std                    1.081209          0.852077           7.780782   
min                    1.000000          0.000000           0.000000   
25%                    2.000000          0.000000           6.000000   
50%                    3.000000          1.000000          10.000000   
75%                    4.000000          1.000000          15.000000   
max                    4.000000          3.000000          40.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count            1470.000000      1470.000000     1470.000000   
mean                2.799320         2.761224        7.008163   
std                 1.289271         0.706476        6.126525   
min                 0.000000         1.000000        0.000000   
25%                 2.000000         2.000000        3.000000   
50%                 3.000000         3.000000        5.000000   
75%                 3.000000         3.000000        9.000000   
max                 6.000000         4.000000       40.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count         1470.000000              1470.000000           1470.000000  
mean             4.229252                 2.187755              4.123129  
std              3.623137                 3.222430              3.568136  
min              0.000000                 0.000000              0.000000  
25%              2.000000                 0.000000              2.000000  
50%              3.000000                 1.000000              3.000000  
75%              7.000000                 3.000000              7.000000  
max             18.000000                15.000000             17.000000  

[8 rows x 23 columns]

In [14]:
descriptive = pd.DataFrame(index = ["Mean", "Median", "Mode", "Q1:25%", "Q2:50%", "Q3:75%", "85%", "Q4:100%", "IQR", "1.5 rule", 
                                    "Lesser", "Greater", "Min", "Max", "Skewness", "Kurtosis"], columns = quan)
for columnname in quan:
    descriptive[columnname]["Mean"] = dataset[columnname].mean() 
    descriptive[columnname]["Median"] = dataset[columnname].median()
    descriptive[columnname]["Mode"] = dataset[columnname].mode()[0]
    descriptive[columnname]["Q1:25%"] = dataset.describe()[columnname]["25%"]
    descriptive[columnname]["Q2:50%"] = dataset.describe()[columnname]["50%"]
    descriptive[columnname]["Q3:75%"] = dataset.describe()[columnname]["75%"]
    descriptive[columnname]["85%"] = np.percentile(dataset[columnname],85)
    descriptive[columnname]["Q4:100%"] = dataset.describe()[columnname]["max"]
    descriptive[columnname]["IQR"] = descriptive[columnname]["Q3:75%"] - descriptive[columnname]["Q1:25%"]
    descriptive[columnname]["1.5 rule"] = 1.5 * descriptive[columnname]["IQR"]
    descriptive[columnname]["Lesser"] =  descriptive[columnname]["Q1:25%"] - descriptive[columnname]["1.5 rule"] 
    descriptive[columnname]["Greater"] = descriptive[columnname]["Q3:75%"] + descriptive[columnname]["1.5 rule"]
    descriptive[columnname]["Min"] = descriptive[columnname].min()
    descriptive[columnname]["Max"] = descriptive[columnname].max()
    descriptive[columnname]["Skewness"] = descriptive[columnname].skew()
    descriptive[columnname]["Kurtosis"] = descriptive[columnname].kurtosis()

In [15]:
descriptive

Age   DailyRate DistanceFromHome Education  \
Mean      36.92381  802.485714         9.192517  2.912925   
Median        36.0       802.0              7.0       3.0   
Mode            35         691                2         3   
Q1:25%        30.0       465.0              2.0       2.0   
Q2:50%        36.0       802.0              7.0       3.0   
Q3:75%        43.0      1157.0             14.0       4.0   
85%           47.0      1297.0             20.0       4.0   
Q4:100%       60.0      1499.0             29.0       5.0   
IQR           13.0       692.0             12.0       2.0   
1.5 rule      19.5      1038.0             18.0       3.0   
Lesser        10.5      -573.0            -16.0      -1.0   
Greater       62.5      2195.0             32.0       7.0   
Min           10.5      -573.0            -16.0      -1.0   
Max           62.5      2195.0             32.0       7.0   
Skewness  0.061801   -0.275424         -0.36695 -0.148446   
Kurtosis -0.926893    0.044548        -0.334971 -0.107945   

         EnvironmentSatisfaction HourlyRate JobInvolvement  JobLevel  \
Mean                    2.721769  65.891156       2.729932  2.063946   
Median                       3.0       66.0            3.0       2.0   
Mode                           3         66              3         1   
Q1:25%                       2.0       48.0            2.0       1.0   
Q2:50%                       3.0       66.0            3.0       2.0   
Q3:75%                       4.0      83.75            3.0       3.0   
85%                          4.0       90.0            3.0       3.0   
Q4:100%                      4.0      100.0            4.0       5.0   
IQR                          2.0      35.75            1.0       2.0   
1.5 rule                     3.0     53.625            1.5       3.0   
Lesser                      -1.0     -5.625            0.5      -2.0   
Greater                      7.0    137.375            4.5       6.0   
Min                         -1.0     -5.625            0.5      -2.0   
Max                          7.0    137.375            4.5       6.0   
Skewness               -0.071139  -0.066863      -0.250481 -0.245159   
Kurtosis                0.153084  -0.365373      -0.845389 -0.206731   

         JobSatisfaction MonthlyIncome  ... PerformanceRating  \
Mean            2.728571   6502.931293  ...          3.153741   
Median               3.0        4919.0  ...               3.0   
Mode                   4          2342  ...                 3   
Q1:25%               2.0        2911.0  ...               3.0   
Q2:50%               3.0        4919.0  ...               3.0   
Q3:75%               4.0        8379.0  ...               3.0   
85%                  4.0       10927.8  ...               4.0   
Q4:100%              4.0       19999.0  ...               4.0   
IQR                  2.0        5468.0  ...               0.0   
1.5 rule             3.0        8202.0  ...               0.0   
Lesser              -1.0       -5291.0  ...               3.0   
Greater              7.0       16581.0  ...               3.0   
Min                 -1.0       -5291.0  ...               0.0   
Max                  7.0       19999.0  ...               4.0   
Skewness       -0.172342      0.184557  ...         -1.231885   
Kurtosis          0.0143     -0.284621  ...         -0.284853   

         RelationshipSatisfaction StockOptionLevel TotalWorkingYears  \
Mean                     2.712245         0.793878         11.279592   
Median                        3.0              1.0              10.0   
Mode                            3                0                10   
Q1:25%                        2.0              0.0               6.0   
Q2:50%                        3.0              1.0              10.0   
Q3:75%                        4.0              1.0              15.0   
85%                           4.0              2.0              20.0   
Q4:100%                       4.0              3.0              40.0  

#### Finding Outliers

In [16]:
def FindOutliers(quan):
    lesser = []
    greater = []
    for columnname in quan:
        if descriptive[columnname]["Min"] < descriptive[columnname]["Lesser"]:
            lesser.append(columnname)
        if descriptive[columnname]["Max"] > descriptive[columnname]["Greater"]:
            greater.append(columnname)        
    return lesser, greater

In [17]:
lesser, greater = FindOutliers(quan)

In [18]:
print("Lesser:", lesser)
print("\nGreater:", greater)

Lesser: ['PerformanceRating']

Greater: ['MonthlyIncome', 'NumCompaniesWorked', 'PerformanceRating', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']


#### Replace Outliers

In [19]:
def replaceOutliers():
    for columnname in lesser:
        dataset[columnname][dataset[columnname] < descriptive[columnname]["Lesser"]] = descriptive[columnname]["Lesser"]
    for columnname in greater:
        dataset[columnname][dataset[columnname] > descriptive[columnname]["Greater"]] = descriptive[columnname]["Greater"]
    return lesser, greater

In [20]:
lesser, greater = replaceOutliers()

In [21]:
lesser

['PerformanceRating']

In [22]:
greater

['MonthlyIncome',
 'NumCompaniesWorked',
 'PerformanceRating',
 'StockOptionLevel',
 'TotalWorkingYears',
 'TrainingTimesLastYear',
 'YearsAtCompany',
 'YearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager']

In [23]:
descriptive = pd.DataFrame(index = ["Mean", "Median", "Mode", "Q1:25%", "Q2:50%", "Q3:75%", "85%", "Q4:100%", "IQR", "1.5 rule", 
                                    "Lesser", "Greater", "Min", "Max", "Skewness", "Kurtosis"], columns = quan)
for columnname in quan:
    descriptive[columnname]["Mean"] = dataset[columnname].mean() 
    descriptive[columnname]["Median"] = dataset[columnname].median()
    descriptive[columnname]["Mode"] = dataset[columnname].mode()[0]
    descriptive[columnname]["Q1:25%"] = dataset.describe()[columnname]["25%"]
    descriptive[columnname]["Q2:50%"] = dataset.describe()[columnname]["50%"]
    descriptive[columnname]["Q3:75%"] = dataset.describe()[columnname]["75%"]
    descriptive[columnname]["85%"] = np.percentile(dataset[columnname],85)
    descriptive[columnname]["Q4:100%"] = dataset.describe()[columnname]["max"]
    descriptive[columnname]["IQR"] = descriptive[columnname]["Q3:75%"] - descriptive[columnname]["Q1:25%"]
    descriptive[columnname]["1.5 rule"] = 1.5 * descriptive[columnname]["IQR"]
    descriptive[columnname]["Lesser"] =  descriptive[columnname]["Q1:25%"] - descriptive[columnname]["1.5 rule"] 
    descriptive[columnname]["Greater"] = descriptive[columnname]["Q3:75%"] + descriptive[columnname]["1.5 rule"]
    descriptive[columnname]["Min"] = descriptive[columnname].min()
    descriptive[columnname]["Max"] = descriptive[columnname].max()
    descriptive[columnname]["Skewness"] = descriptive[columnname].skew()
    descriptive[columnname]["Kurtosis"] = descriptive[columnname].kurtosis()

In [24]:
descriptive

Age   DailyRate DistanceFromHome Education  \
Mean      36.92381  802.485714         9.192517  2.912925   
Median        36.0       802.0              7.0       3.0   
Mode            35         691                2         3   
Q1:25%        30.0       465.0              2.0       2.0   
Q2:50%        36.0       802.0              7.0       3.0   
Q3:75%        43.0      1157.0             14.0       4.0   
85%           47.0      1297.0             20.0       4.0   
Q4:100%       60.0      1499.0             29.0       5.0   
IQR           13.0       692.0             12.0       2.0   
1.5 rule      19.5      1038.0             18.0       3.0   
Lesser        10.5      -573.0            -16.0      -1.0   
Greater       62.5      2195.0             32.0       7.0   
Min           10.5      -573.0            -16.0      -1.0   
Max           62.5      2195.0             32.0       7.0   
Skewness  0.061801   -0.275424         -0.36695 -0.148446   
Kurtosis -0.926893    0.044548        -0.334971 -0.107945   

         EnvironmentSatisfaction HourlyRate JobInvolvement  JobLevel  \
Mean                    2.721769  65.891156       2.729932  2.063946   
Median                       3.0       66.0            3.0       2.0   
Mode                           3         66              3         1   
Q1:25%                       2.0       48.0            2.0       1.0   
Q2:50%                       3.0       66.0            3.0       2.0   
Q3:75%                       4.0      83.75            3.0       3.0   
85%                          4.0       90.0            3.0       3.0   
Q4:100%                      4.0      100.0            4.0       5.0   
IQR                          2.0      35.75            1.0       2.0   
1.5 rule                     3.0     53.625            1.5       3.0   
Lesser                      -1.0     -5.625            0.5      -2.0   
Greater                      7.0    137.375            4.5       6.0   
Min                         -1.0     -5.625            0.5      -2.0   
Max                          7.0    137.375            4.5       6.0   
Skewness               -0.071139  -0.066863      -0.250481 -0.245159   
Kurtosis                0.153084  -0.365373      -0.845389 -0.206731   

         JobSatisfaction MonthlyIncome  ... PerformanceRating  \
Mean            2.728571   6361.891837  ...               3.0   
Median               3.0        4919.0  ...               3.0   
Mode                   4         16581  ...                 3   
Q1:25%               2.0        2911.0  ...               3.0   
Q2:50%               3.0        4919.0  ...               3.0   
Q3:75%               4.0        8379.0  ...               3.0   
85%                  4.0       10927.8  ...               3.0   
Q4:100%              4.0       16581.0  ...               3.0   
IQR                  2.0        5468.0  ...               0.0   
1.5 rule             3.0        8202.0  ...               0.0   
Lesser              -1.0       -5291.0  ...               3.0   
Greater              7.0       16581.0  ...               3.0   
Min                 -1.0       -5291.0  ...               0.0   
Max                  7.0         16581  ...               3.0   
Skewness       -0.172342     -0.382944  ...         -1.565624   
Kurtosis          0.0143     -0.715593  ...          0.398369   

         RelationshipSatisfaction StockOptionLevel TotalWorkingYears  \
Mean                     2.712245         0.764966         11.103061   
Median                        3.0              1.0              10.0   
Mode                            3              0.0              10.0   
Q1:25%                        2.0              0.0               6.0   
Q2:50%                        3.0              1.0              10.0   
Q3:75%                        4.0              1.0              15.0   
85%                           4.0              2.0              20.0   
Q4:100%                       4.0              2.5              28.5  

In [25]:
# IQR = 0 So treating them as no outlier column
# Skip columns if it has no variation

def FindOutliers(quan):
    lesser = []
    greater = []
    for columnname in quan:
        if descriptive[columnname]["IQR"] == 0:
           continue
        if descriptive[columnname]["Min"] < descriptive[columnname]["Lesser"]:
            lesser.append(columnname)
        if descriptive[columnname]["Max"] > descriptive[columnname]["Greater"]:
            greater.append(columnname)        
    return lesser, greater

In [26]:
lesser, greater = FindOutliers(quan)

In [27]:
lesser

[]

In [28]:
greater

[]

#### 1. How many Employees are highly job satisfied?

In [29]:
def freqtable(columnname, dataset):
    freqtable = pd.DataFrame(columns = ["Unique Values", "Frequency", "Relative Frequency", "Cumulative Frequency"])

    freqtable["Unique Values"] = dataset[columnname].value_counts().index
    freqtable["Frequency"] = dataset[columnname].value_counts().values
    freqtable["Relative Frequency"] = freqtable["Frequency"]/1470
    freqtable["Cumulative Frequency"] = dataset[columnname].cumsum()
    return freqtable

In [30]:
freqtable("JobSatisfaction", dataset)

Unique Values  Frequency  Relative Frequency  Cumulative Frequency
0              4        459            0.312245                     4
1              3        442            0.300680                     6
2              1        289            0.196599                     9
3              2        280            0.190476                    12

Out of 1470 Employees, 459 Employees are highly satisfied with their job. So they given the rating of 4.
569 Employees are not really satisfied with their job.

#### How many employees had a overall good worklife balance?

In [31]:
freqtable("WorkLifeBalance", dataset)

Unique Values  Frequency  Relative Frequency  Cumulative Frequency
0              3        893            0.607483                     1
1              2        344            0.234014                     4
2              4        153            0.104082                     7
3              1         80            0.054422                    10

Out of 1470 Employees, 153 had a good work life balance and 
893 had a moderate work life balance. 
424 employees are not really experience a good work life balance. 

#### How many Male and Female Employees?

In [32]:
dataset['Gender'].value_counts()

Gender
Male      882
Female    588
Name: count, dtype: int64

#### What is the average salary of a Male Employer?

In [33]:
male_avg = dataset[dataset["Gender"].str.contains('Male')]
Avg_male_salary = male_avg['MonthlyIncome'].mean()
print(Avg_male_salary)

6225.730158730159


In [34]:
male_avg = dataset[dataset["Gender"].str.contains('Male')]
Avg_male_salary = male_avg.groupby(['MonthlyIncome','MonthlyRate'])['PercentSalaryHike'].mean().reset_index()
print(Avg_male_salary.head(10))

   MonthlyIncome  MonthlyRate  PercentSalaryHike
0           1009        26999               11.0
1           1051        13493               15.0
2           1052        23384               22.0
3           1081        16019               13.0
4           1091        10642               17.0
5           1102         9241               22.0
6           1118         8040               14.0
7           1223        16901               22.0
8           1232        19281               14.0
9           1274         7152               13.0


So Average Salary of Male is 6225 with average monthly rate can assume like incentive of 26999. 
And Maximum Salary hike would be 22%

#### What is the Average salary of a Female Employer?

In [42]:
female_avg = dataset[dataset['Gender'].str.contains('Female')]
print(f'Average monthly salary of female = {female_avg['MonthlyIncome'].mean()}')

Avg_female_salary = female_avg.groupby(['MonthlyIncome', 'MonthlyRate'])['PercentSalaryHike'].mean().reset_index()
print(Avg_female_salary.head(10))

Average monthly salary of female = 6566.134353741497
   MonthlyIncome  MonthlyRate  PercentSalaryHike
0           1129        17536               11.0
1           1200         9724               12.0
2           1261        22262               12.0
3           1359        16154               12.0
4           1416        17258               13.0
5           1483        16102               14.0
6           1514         8018               16.0
7           1569        18420               12.0
8           1611        19305               15.0
9           1859         6148               25.0


So Average monthly salary for female is 6566 and maximum monthly rate is 22262. And Maximum hike for about 25%

In [48]:
from scipy.stats import ttest_ind, chi2_contingency

# Make a copy of dataset
df = dataset.copy()

# Identify target
target = "Attrition"   # (values: "Yes", "No")

# Split groups
Attrition_yes = df[df[target] == "Yes"]
Attrition_no = df[df[target] == "No"]

print("===== Numerical Features (t-test) =====")
for col in quan:
    t_stat, p_val = ttest_ind(Attrition_yes[col], Attrition_no[col])
    result = "Significant" if p_val < 0.05 else "Not Significant"
    print(f"{col}: p={p_val:.4f} → {result}")

print("\n===== Categorical Features (Chi-square) =====")
for col in qual:
    contingency = pd.crosstab(df[col], df[target])
    chi2, p, dof, expected = chi2_contingency(contingency)
    result = "Significant" if p < 0.05 else "Not Significant"
    print(f"{col}: p={p:.4f} → {result}")

===== Numerical Features (t-test) =====
Age: p=0.0000 → Significant
DailyRate: p=0.0299 → Significant
DistanceFromHome: p=0.0028 → Significant
Education: p=0.2293 → Not Significant
EnvironmentSatisfaction: p=0.0001 → Significant
HourlyRate: p=0.7931 → Not Significant
JobInvolvement: p=0.0000 → Significant
JobLevel: p=0.0000 → Significant
JobSatisfaction: p=0.0001 → Significant
MonthlyIncome: p=0.0000 → Significant
MonthlyRate: p=0.5611 → Not Significant
NumCompaniesWorked: p=0.1002 → Not Significant
PercentSalaryHike: p=0.6056 → Not Significant
PerformanceRating: p=nan → Not Significant
RelationshipSatisfaction: p=0.0787 → Not Significant
StockOptionLevel: p=0.0000 → Significant
TotalWorkingYears: p=0.0000 → Significant
TrainingTimesLastYear: p=0.0349 → Significant
WorkLifeBalance: p=0.0142 → Significant
YearsAtCompany: p=0.0000 → Significant
YearsInCurrentRole: p=0.0000 → Significant
YearsSinceLastPromotion: p=0.1509 → Not Significant
YearsWithCurrManager: p=0.0000 → Significant

====